### PR TITLE
Whole new editor

### PR DESCRIPTION
**Pull request by farskid**
### Changes in PR:

- Adds Monaco editor with support for code completion and IntelliSense for xstate types (no more typos!)
- Adds `react-app-rewired` to extend CRA Webpack config without ejecting (this is needed for Monaco workers)
- Adds prettier autoformatting by hitting CTRL (CMD) + S on the editor
- Updates visualizer by hitting CTRL (CMD) + R on the editor

### Hacks in PR:

- To enable IntelliSense for xstate typings, we need to access declaration files. I simply copied the latest version declarations, but this ought to be automatic on xstate releases (can also be automated on dependancy updates based on the xstate version we install in this repo).

### Concerns

- There was a CSS rule in `public/App.css`: `*, *:before, *:after {positin: relative;}` which is inherited by some critical elements on the page such as event triggers in the viz or show/hide button in the header. This breaks hint tooltips of Monaco. I needed to add an override to get around this for now, but who knows what else this global CSS breaks?

- Also since the above CSS rule is inside `public` directory, whoever installs the xstate-viz package locally, won't get the effects of that rule. for instance, clicking on event triggers won't work locally! (tested)

- Having to add Monaco Webpack plugin for web workers in the `public` repo was not a choice! there is *no* support for other bundlers for now. this is a concern of mine since it breaks portability of viz package. people need to use it locally inside a Webpack project in order to properly load the workers. local installs might not be a focus for now but are good to have in mind. this also means we need to document how to set up the editor locally for those unfortunate users.

### CODE REVIEW

File: src/Editor.tsx [hunk 1]
```diff
@@ -49,18 +53,28 @@ export const Editor: React.FunctionComponent<EditorProps> = props => {
...
+        registerEditorActions={[
+          // Register CTRL+R (CMD + R) to update the visualizer
+          {
+            id: 'update-viz',
+            label: 'Update Visualizer',
+            keybindings: [
+              MonacoEditor.KeyMod.CtrlCmd | MonacoEditor.KeyCode.KEY_R
+            ],
+            run: ed => {
+              console.log('update viz');
```
Comment (davidkpiano) [inline src/Editor.tsx]:
```suggestion
```

File: src/MonacoEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  let subscription: monaco.IDisposable;
+  const editorRef = useRef<HTMLDivElement>(null!);
+  let editor: monaco.editor.IStandaloneCodeEditor;
+
+  /**
+   * In case editor is controlled in future, this needs to be changed to `useLayoutEffect`
+   * to avoid stale values considering React batches state updates and effect execution
+   */
+  useEffect(() => {
+    console.log('running effect');
```
Comment (davidkpiano) [inline src/MonacoEditor.tsx]:
```suggestion
```

File: src/MonacoEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    // Run format on initial code
+    editor.getAction('editor.action.formatDocument').run();
+
+    // Subscribe to editor model content change
+    subscription = editor.onDidChangeModelContent(() => {
+      props.onChange(editor.getValue());
+    });
+
+    return () => {
+      console.log('cleanup');
```
Comment (davidkpiano) [inline src/MonacoEditor.tsx]:
```suggestion
```

File: src/MonacoEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+      if (editor) {
+        console.log('cleaning editor');
```
Comment (davidkpiano) [inline src/MonacoEditor.tsx]:
```suggestion
```

File: src/MonacoEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        (editor.getModel() as monaco.IDisposable).dispose();
+        editor.dispose();
+      }
+      if (subscription) {
+        console.log('cleaning subscription');
```
Comment (davidkpiano) [inline src/MonacoEditor.tsx]:
```suggestion
```

**Comment by farskid** [comment]
@davidkpiano Do you know of a good way to maintain xstate definitions for the editor?

**Comment by davidkpiano** [comment]
> @davidkpiano Do you know of a good way to maintain xstate definitions for the editor?

We'll keep this way for now and make some script to copy over definitions.

**Comment by Andarist** [comment]
Have you considered using [Sandpack](https://github.com/codesandbox/codesandbox-client/tree/master/standalone-packages/sandpack)? Would allow for some neat things - not sure at the moment though how this would fit into shareable gists etc right now.

**Comment by farskid** [comment]
Sandpack looks good, maybe for a bit later.
image

**Comment by Andarist** [comment]
😅 I actually thought it's more ready and that CodeSandbox is using this in their own implementation by now - doesn't seem to be the case though. Just remembered that Babel REPL was gonna be reimplemented on top of that, but I see now that this effort got stale as well 😢

**Comment by farskid** [comment]
@davidkpiano ping

**Pull request merged**